Sources/Basics/Observability.swift

```diff
@@ -0,0 +1,400 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2021 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+ */
+
+import TSCBasic
+import TSCUtility
+
+typealias TSCDiagnostic = TSCBasic.Diagnostic
+
+// this could become a struct when we remove the "errorsReported" pattern
+
+// designed after https://github.com/apple/swift-log
+// designed after https://github.com/apple/swift-metrics
+public class ObservabilitySystem {
+
+    // global
+
+    private static var _global = ObservabilitySystem(factory: NOOPFactory())
+    private static var bootstrapped = false
+    private static let lock = Lock()
+
+    public static func bootstrapGlobal(factory: ObservabilityFactory) {
+        Self.lock.withLock {
+            // FIXME: disabled for testing
+            //precondition(!Self.bootstrapped, "ObservabilitySystem can only bootstrapped once")
+            Self._global = .init(factory: factory)
+            Self.bootstrapped = true
+        }
+    }
+
+    @available(*, deprecated, message: "this pattern is deprecated, transition to error handling instead")
+    public static var errorsReported: Bool {
+        Self.lock.withLock {
+            Self._global.errorsReported
+        }
+    }
+
+    // as we transition to async/await we can take advantage of Task Local values instead of a global
```

Comment:
Is there a top-level Task for the main entry point?  If not, then it would still make sense to have a fall-back for diagnostics emitted outside of a task.

Reply:
agreed